### ISSUE TITLE
Synchronous view model initialization

### ISSUE THREAD

**Issue by korzonkiee** [MvvmCross/MvvmCross#1902]
Hey,

Before I start, I would like to mention that I'm using new `NavigationService` introduced in MvvmCross 5.0.0.

Now, let's say I have two view models, `FirstViewModel` and `SecondViewModel`, and I want to pass data from `FirstViewModel` to `SecondViewModel`.

What I would like to achieve is to receive data in SecondViewModel and use that data inside view model's `Appeared()` method. (In my particular case, I'm passing push notification data received in Android's `MainLauncher` Activity (when an application was in the background) straight to the `MainViewModel`, where I would like to show user dialog as soon as the `MainViewModel` has appeared.)

So, inside `FirstViewModel` I'm navigating to `SecondViewModel` as follows:

```
var param = new NavParam("some_text");
await navigationService.Navigate<SecondViewModel, NavParam>(param);
```

And receive data in `SecondViewModel`:
```
private string parameterValue;

public SecondViewModel()
{
    // ...
}

public override Task Initialize(NavParam parameter)
{
    await Task.Run(() => parameterValue = parameter.Value);
}

public override void Appeared()
{
    base.Appeared();

    if (parameterValue != null)
    {
        // do something with data (for example show pop-up)
    }
}
```

The problem is that, since ViewModel's `Initialize()` method is asynchronous, the `parameterValue` inside `Appeared()` method may not be initialized.

I'm looking for some way to either:
- initialize view model synchronously (formerly we could synchronously initialize view model with data using `ShowViewModel<SecondViewModel>(new { index = 1 })` and `Init(int index)` and use received that inside `Appeared()` knowing that is has been initialized,
or
- perform some synchronization, like awaiting initialization process inside `Appeared()` method.

I'm pretty sure that I'm missing something, but if you could point me to the right direction, I would be grateful.

By the way, I've been investigating this issue and I've noticed some strange behaviour, namely, sometimes, even though the `Initialize()` process finished his work before `Appeared()`, the `parameterValue` was still uninitialized.

Besides that, I've noticed that, while code runs inside `Initialize()` method, then the reference to the ViewModel is not the same as the reference to the ViewModel while code runs inside `Appeared()` method.

Moreover, when I've introduced an `static int instanceCounter` variable in order to count number of times that the view model is being created:
```
private static int instanceCounter = 0;

public SecondViewModel()
{
    instanceCounter++;
}
```

it came out that the view model is constructed twice.

I don't know if it's somehow related to my problem, but I wanted to share my observation and want you to take a look at those issues. Maybe I'm missing something obvious.

Thanks

**Version:** 5.0.1

**Platform:** 
- [ ] Android

**Comment by Bowman74** [comment]
This is likely related to #1905 that I just entered. @korzonkiee is right that two instances of the view model are likely being created. Like the UWP issue, the instance that the view has a reference to isn't the same instance that has the Initialize method called on it.

**Comment by cstoltze** [comment]
I also encountered this. Here is what I learned about the navigation service in 5.0.0 and 5.0.1 on Android.

I used the navigation service to navigate to a view model using:
`_navigationService.Navigate<SecondViewModel>();`

The SecondViewModel tracks variable values and the order of events in its construction and initialization by using `Debug.WriteLine`. Here is the code for the SecondViewModel:



    public class SecondViewModel
        : MvxViewModel
    {
        // Strings that monitor the state of the ViewModel
        private string _constructionString = "Not Yet Set";
        private string _initializationString = "Not Yet Set";

        public SecondViewModel()
        {
            _constructionString = "Set in Constructor";

            WriteDebugValues("Constructing");
        }

        public override Task Initialize()
        {
            _initializationString = "Set in Initialization";

            WriteDebugValues("Initializing");

            return base.Initialize();
        }

        private IMvxCommand _writeDebugValuesCommand;
        public IMvxCommand WriteDebugValuesCommand
        {
            get
            {
                _writeDebugValuesCommand = _writeDebugValuesCommand ?? new MvxCommand(() => WriteDebugValues("Button Clicked"));
                return _writeDebugValuesCommand;
            }
        }

        public void WriteDebugValues(string title)
        {
            Debug.WriteLine(title);
            Debug.WriteLine("_constructionString = {0}", _constructionString);
            Debug.WriteLine("_initializationString = {0}", _initializationString);
        }
    }




I tested navigation to the SecondViewModel in both MvvmCross 5.0.0 and 5.0.1 on Android.

5.0.0 exhibits the expected behavior.
In 5.0.0, when navigating to the SecondViewModel, the debugger displayed:

[0:] Constructing
[0:] _constructionString = Set in Constructor
[0:] _initializationString = Not Yet Set
[0:] Initializing
[0:] _constructionString = Set in Constructor
[0:] _initializationString = Set in Initialization
06-02 07:50:47.164 I/mvx     ( 3577):   4.17 Embedded ViewModel used
[0:] mvx:Diagnostic:  4.17 Embedded ViewModel used
06-02 07:50:47.480 D/OpenGLRenderer( 3577): endAllStagingAnimators on 0xaf53e580 (RippleDrawable) with handle 0xb0266210

Upon clicking the button bound to the `WriteDebugValuesCommand`, the debugger displayed:

[0:] Button Clicked
[0:] _constructionString = Set in Constructor
[0:] _initializationString = Set in Initialization

5.0.1 does NOT exhibit the expected behavior. It appears to load a new ViewModel from Intent with Extras after a ViewModel has already been constructed and initialized. This new ViewModel is only constructed; it is NOT initialized. See details below:

In 5.0.1, when navigating to the SecondViewModel, the debugger displayed:

[0:] Constructing
[0:] _constructionString = Set in Constructor
[0:] _initializationString = Not Yet Set
[0:] Initializing
[0:] _constructionString = Set in Constructor
[0:] _initializationString = Set in Initialization
06-02 07:56:24.510 I/mvx     ( 3892):   3.66 Attempting to load new ViewModel from Intent with Extras
[0:] mvx:Diagnostic:  3.66 Attempting to load new ViewModel from Intent with Extras
[0:] Constructing
[0:] _constructionString = Set in Constructor
[0:] _initializationString = Not Yet Set
06-02 07:56:24.970 D/OpenGLRenderer( 3892): endAllStagingAnimators on 0xaf53e580 (RippleDrawable) with handle 0xb0266220

Upon clicking the button bound to the `WriteDebugValuesCommand`, the debugger displayed:

[0:] Button Clicked
[0:] _constructionString = Set in Constructor
[0:] _initializationString = Not Yet Set


I hope that is helpful information for you all. Thank you for the work you put into MvvmCross.
